// abstract-things/sensors/index.js
'use strict';

module.exports.Sensor = require('./sensor');

module.exports.AtmosphericPressure = require('./atmospheric-pressure');
module.exports.CarbonDioxideDetection = require('./carbon-dioxide-detection');
module.exports.CarbonDioxideLevel = require('./carbon-dioxide-level');
module.exports.CarbonMonoxideDetection = require('./carbon-monoxide-detection');
module.exports.CarbonMonoxideLevel = require('./carbon-monoxide-level');
module.exports.Contact = require('./contact');
module.exports.Illuminance = require('./illuminance');
module.exports.Motion = require('./motion');
module.exports.PM2_5 = require('./pm2_5');
module.exports.PM10 = require('./pm10');
module.exports.PowerConsumed = require('./power-consumed');
module.exports.PowerLoad = require('./power-load');
module.exports.RelativeHumidity = require('./relative-humidity');
module.exports.Temperature = require('./temperature');
module.exports.SmokeDetection = require('./smoke-detection');
module.exports.Voltage = require('./voltage');
module.exports.WaterDetection = require('./water-detection');

module.exports.Depth = require('./depth');
module.exports.Density = require('./density');
