// mimorelinks/libs/GatewayHelper.js
const Gateway = require("./Gateway");
const utils = require("./utils");

class GatewayHelper {
  constructor(platform) {
    if (!platform) {
      throw new Error("[error] [mimorelink] [GatewayHelper:constructor] Param error");
    }
    this.gateways = {};
    this.platform = platform;
  }

  add(gateway) {
    if (!gateway || !gateway.sid || this.gateways.hasOwnProperty(gateway.sid)) {
      return;
    }
    this.gateways[gateway.sid] = gateway;
  }

  addOrUpdate(data) {
    let sid = data.sid;
    if (this.gateways.hasOwnProperty(sid)) {
      let gateway = this.gateways[sid];
      gateway.update(data);
    } else {
      this.gateways[sid] = new Gateway(data);
    }
  }

  remove(sid) {
    delete this.gateways[sid];
  }

  getBySid(sid) {
    return this.gateways[sid];
  }

  uploadBySid(sid, data) {
    if (this.gateways.hasOwnProperty(sid)) {
      let gateway = this.gateways[sid];
      gateway.update(data);
    }
  }

  getGatewayList() {
    let list = [];
    for (let key in this.gateways) {
      let value = this.gateways[key];
      list.push(value);
    }
    return list;
  }

  getAll() {
    return this.gateways;
  }

  getIdList(sid) {
    let gateway = this.getBySid(sid);
    if (gateway) {
      this.platform.send(gateway.ip, gateway.port, {
        cmd: "get_id_list"
      });
    } else {
      console.error("[error] [mimorelink] [GatewayHelper:getIdList] sid:%s is not exit", sid);
    }
  }

  read(sid) {
    let gateway = this.getBySid(sid);
    if (gateway) {
      this.platform.send(gateway.ip, gateway.port, {
        cmd: "read",
        sid: sid
      });
    } else {
      console.error("[error] [mimorelink] [GatewayHelper:read] sid:%s can not find gateway", sid);
    }
  }

  write(sid, data) {
    // console.log("[log] [mimorelink] [GatewayHelper:write] sid=%s", sid);
    // console.log("[log] [mimorelink] [GatewayHelper:write] data=%s", JSON.stringify(data));
    let gateway = this.getBySid(sid);
    if (gateway) {
      let msg = {
        cmd: "write",
        model: "gateway",
        sid: gateway.sid,
        data: Object.assign({}, data)
      };
      msg.data.key = utils.cipher(gateway.token, gateway.password, gateway.iv);
      this.platform.send(gateway.ip, gateway.port, msg);
    } else {
      console.error("[error] [mimorelink] [GatewayHelper:read] sid:%s can not find", sid);
    }
  }

  controlLightPower({ sid, power }) {
    let prepValue = 0;
    let hue = 0,
      saturation = 0,
      brightness = 0;
    if (power) {
      if (!hue) {
        hue = 0;
      }
      if (!saturation) {
        saturation = 0 * 100;
      }
      if (!brightness) {
        brightness = 50;
      }
      let rgb = utils.hsb2rgb([hue, saturation / 100, 1]);
      prepValue = parseInt(utils.dec2hex(brightness, 2) + utils.dec2hex(rgb[0], 2) + utils.dec2hex(rgb[1], 2) + utils.dec2hex(rgb[2], 2), 16);
    }

    this.write(sid, { rgb: prepValue });
  }

  controlLightHLS({ sid, hue, saturation, brightness }) {
    let prepValue = 0;

    if (hue == 0 && saturation == 0 && brightness == 0) {
      this.write(sid, { rgb: prepValue });
    } else {
      let rgb = utils.hsb2rgb([hue, saturation / 100, 1]);
      prepValue = parseInt(utils.dec2hex(brightness, 2) + utils.dec2hex(rgb[0], 2) + utils.dec2hex(rgb[1], 2) + utils.dec2hex(rgb[2], 2), 16);

      this.write(sid, { rgb: prepValue });
    }
  }

  controlLightRGB({ sid, rgb }) {
    let prepValue = 0;
    if (rgb) {
      prepValue = rgb;
    }
    this.write(sid, { rgb: prepValue });
  }

  controlMid({ sid, mid, vol }) {
    let prepValue = 1000;
    let defaultVolume = 50;
    if (mid) {
      prepValue = parseInt(mid);
    }
    if (vol) {
      defaultVolume = parseInt(vol);
    }
    this.write(sid, { mid: prepValue, vol: defaultVolume });
  }
}

module.exports = GatewayHelper;
