// mimorelinks/libs/MiAqara.js
let dgram = require("dgram");
const DeviceHelper = require("mimorelinks/libs/DeviceHelper");
const GatewayHelper = require("mimorelinks/libs/GatewayHelper");
const DeviceMapsHelper = require("mimorelinks/libs/DeviceMapsHelper");
const DeviceParser = require("mimorelinks/libs/DeviceParser");
const utils = require("mimorelinks/libs/utils");

const defaultConfig = {
  iv: Buffer.from([0x17, 0x99, 0x6d, 0x09, 0x3d, 0x28, 0xdd, 0xb3, 0xba, 0x69, 0x5a, 0x2e, 0x6f, 0x58, 0x56, 0x2e]),
  multicastAddress: "224.0.0.50",
  multicastPort: 4321,
  serverPort: 9898,
  bindAddress: ""
};

class MiAqara {
  constructor(gateways, opts) {
    opts = opts ? Object.assign({}, defaultConfig, opts) : defaultConfig;
    this.multicastAddress = opts.multicastAddress;
    this.multicastPort = opts.multicastPort;
    this.serverPort = opts.serverPort;
    this.bindAddress = opts.bindAddress;
    this.searchGatewayHandle = null;
    this.searchGatewayInterval = 10000;
    this.readCount = 0;
    this.onReady = opts.onReady;
    this.onMessage = opts.onMessage;
    this.deviceMapsHelper = new DeviceMapsHelper();
    this.gatewayHelper = new GatewayHelper(this);
    this.deviceHelper = new DeviceHelper(this);
    this.parser = DeviceParser;
    this.gatewaysList = gateways;

    if (Array.isArray(gateways)) {
      for (let i = 0; i < gateways.length; i++) {
        this.gatewayHelper.addOrUpdate({
          iv: gateways[i].iv || defaultConfig.iv,
          sid: gateways[i].sid,
          password: gateways[i].password
        });
      }
    } else if (utils.isObject(gateways)) {
      this.gatewayHelper.addOrUpdate({
        iv: gateways.iv || defaultConfig.iv,
        sid: gateways.sid,
        password: gateways.password
      });
    } else {
      throw new Error("Param error");
    }
  }

  start() {
    this.createSocket();
    this.initServerSocket();
    this.sendWhoisCommand();
  }

  stop() {
    if (this.searchGatewayHandle != null) {
      clearInterval(this.searchGatewayHandle);
    }

    this.gatewayHelper.gateways = {};
    this.deviceHelper.devices = {};

    this.serverSocket && this.serverSocket.close();
  }

  createSocket() {
    this.serverSocket = dgram.createSocket({
      type: "udp4",
      reuseAddr: true
    });
  }

  initServerSocket() {
    let serverSocket = this.serverSocket;
    let that = this;

    serverSocket.on("error", function(err) {
      console.error("[error] [mimorelink]  msg - %s, stack - %s\n", err.message, err.stack);
    });

    serverSocket.on("listening", function() {
      console.info(`[log] [mimorelink] Server is listening on port ${that.serverPort}.`);
      if (!that.bindAddress) {
        serverSocket.addMembership(that.multicastAddress);
      } else {
        serverSocket.setMulticastInterface(that.bindAddress);
        serverSocket.addMembership(that.multicastAddress, that.bindAddress);
      }

      that.searchGatewayHandle = setInterval(that.sendWhoisCommand.bind(that), that.searchGatewayInterval);
    });
    serverSocket.on("message", this.parseMessage.bind(this));

    serverSocket.bind({ port: that.serverPort, exclusive: true });
  }

  parseMessage(msg, rinfo) {
    let data;
    try {
      data = JSON.parse(msg);
      if (data.hasOwnProperty("data")) {
        data.data = JSON.parse(data.data);
      }
    } catch (e) {
      console.error("[error] [mimorelink] Bad message: %s", msg);
      return;
    }
    let cmd = data["cmd"];

    if (cmd === "iam") {
      for (let i = 0; i < this.gatewaysList.length; i++) {
        if (this.gatewaysList[i].sid == data.sid) {
          this.gatewayHelper.uploadBySid(data.sid, data);
          this.gatewayHelper.getIdList(data.sid);
          this.gatewayHelper.read(data.sid);
        }
      }
    } else if (cmd === "get_id_list_ack") {
      for (let i = 0; i < this.gatewaysList.length; i++) {
        if (this.gatewaysList[i].sid == data.sid) {
          this.gatewayHelper.uploadBySid(data.sid, data);
          this.deviceMapsHelper.addOrUpdate(data.sid, data.data);
          this.deviceHelper.readAll(data.data);
          this.readCount += data.data.length;
        }
      }
    } else if (cmd === "report") {
      this._addOrUpdate(data);
    } else if (cmd === "read_ack") {
      this._addOrUpdate(data);
      this.readCount--;
      if (this.readCount === 0) {
        this.onReady && this.onReady(data);
      }
    } else if (cmd === "write_ack") {
      this._addOrUpdate(data);
    } else if (cmd === "server_ack") {
    } else if (cmd === "heartbeat") {
      this._addOrUpdate(data);
    }

    this.onMessage && this.onMessage(data);
  }

  _addOrUpdate(data) {
    if (!data) {
      return;
    }
    if (data["model"] === "gateway") {
      for (let i = 0; i < this.gatewaysList.length; i++) {
        if (this.gatewaysList[i].sid == data.sid) {
          this.gatewayHelper.uploadBySid(data.sid, data);
        }
      }
    } else {
      this.deviceHelper.addOrUpdate(data);
    }
  }

  send(ip, port, msg) {
    if (!ip || !port || !msg) {
      throw new Error("Param error");
    }
    let msgStr = utils.messageFormat(msg);
    this.serverSocket.send(msgStr, 0, msgStr.length, port, ip);
  }

  sendWhoisCommand() {
    this.send(this.multicastAddress, this.multicastPort, {
      cmd: "whois"
    });
  }
}

module.exports = MiAqara;
