// mimorelinks/libs/utils.js
const crypto = require("crypto");

module.exports = {
  isObject(obj) {
    return obj && Object.prototype.toString.apply(obj) === "[object Object]";
  },

  messageFormat(msg) {
    try {
      let msgData = Object.assign({}, msg);
      if (this.isObject(msgData.data)) {
        msgData.data = JSON.stringify(msgData.data);
      }
      return JSON.stringify(msgData);
    } catch (e) {
      console.error("[error] [mimorelink] [messageFormat] Bad msg!", msg);
      return "{}";
    }
  },

  cipher(token, password, iv) {
    let cipher = crypto.createCipheriv("aes-128-cbc", password, iv);
    let key = cipher.update(token, "ascii", "hex");
    cipher.final("hex");
    return key;
  },

  hsb2rgb(hsb) {
    let rgb = [];
    for (let offset = 240, i = 0; i < 3; i++, offset -= 120) {
      let x = Math.abs(((hsb[0] + offset) % 360) - 240);
      if (x <= 60) rgb[i] = 255;
      else if (60 < x && x < 120) rgb[i] = (1 - (x - 60) / 60) * 255;
      else rgb[i] = 0;
    }
    for (let i = 0; i < 3; i++) rgb[i] += (255 - rgb[i]) * (1 - hsb[1]);
    for (let i = 0; i < 3; i++) rgb[i] *= hsb[2];
    for (let i = 0; i < 3; i++) rgb[i] = Math.round(rgb[i]);
    return rgb;
  },

  rgb2hsb(rgb) {
    let hsb = [];
    let rearranged = rgb.slice(0);
    let maxIndex = 0,
      minIndex = 0;
    let tmp;
    for (let i = 0; i < 2; i++) {
      for (let j = 0; j < 2 - i; j++) {
        if (rearranged[j] > rearranged[j + 1]) {
          tmp = rearranged[j + 1];
          rearranged[j + 1] = rearranged[j];
          rearranged[j] = tmp;
        }
      }
    }
    for (let i = 0; i < 3; i++) {
      if (rearranged[0] === rgb[i]) minIndex = i;
      if (rearranged[2] === rgb[i]) maxIndex = i;
    }
    hsb[2] = rearranged[2] / 255.0;
    hsb[1] = 1 - rearranged[0] / rearranged[2];
    hsb[0] = maxIndex * 120 + 60 * (rearranged[1] / hsb[1] / rearranged[2] + (1 - 1 / hsb[1])) * ((maxIndex - minIndex + 3) % 3 === 1 ? 1 : -1);
    hsb[0] = (hsb[0] + 360) % 360;
    return hsb;
  },

  dec2hex(dec, len) {
    let hex = "";
    while (dec) {
      let last = dec & 15;
      hex = String.fromCharCode((last > 9 ? 55 : 48) + last) + hex;
      dec >>= 4;
    }
    if (len) {
      while (hex.length < len) hex = "0" + hex;
    }
    return hex;
  }
};
